### PR TITLE
Add ARM Cortex-M3 DesignStart FPGA reference design support

### PR DESCRIPTION
**Pull request by henrikbrixandersen**
Add support for the ARM Cortex-M3 DesignStart FPGA reference design running on the Digilent Arty development board.

```
$ ./scripts/sanitycheck -p arty_a7_arm_designstart_m3 --device-testing --device-serial /dev/ttyUSB1
INFO    - JOBS: 2
INFO    - Building initial testcase list...
INFO    - 889 test configurations selected, 717 configurations discarded due to filters.

Device testing on:

| Platform                   | ID   | Serial device   |
|----------------------------|------|-----------------|
| arty_a7_arm_designstart_m3 |      | /dev/ttyUSB1    |

INFO    - Adding tasks to the queue...
INFO    - Total complete:  172/ 172  100%  skipped:   18, failed:    0
INFO    - 105 of 105 tests passed (100.00%), 0 failed, 784 skipped with 0 warnings in 1356.04 seconds
INFO    - In total 327 test cases were executed on 1 out of total 282 platforms (0.35%)
INFO    - 105 tests executed on platforms, 0 tests were only built.

Hardware distribution summary:

| Board                      | ID   |   Counter |
|----------------------------|------|-----------|
| arty_a7_arm_designstart_m3 |      |       105 |

```

### CODE REVIEW

File: soc/arm/arm/designstart/soc.h [hunk 1]
```diff
@@ -9,4 +9,6 @@
 
 #include <devicetree.h>
 
+#define __MPU_PRESENT CONFIG_CPU_HAS_ARM_MPU
```
Comment (ioannisg) [inline soc/arm/arm/designstart/soc.h:12]: Where do you define CONFIG_CPU_HAS_ARM_MPU for the Cortex-M3 SoC?

Comment (henrikbrixandersen) [reply soc/arm/arm/designstart/soc.h:12]: It's defined in `boards/arm/arty/Kconfig.board` since the MPU is optional.

Comment (ioannisg) [reply soc/arm/arm/designstart/soc.h:12]: That's why I could not find it 👍 

This is not ideal; CPU_HAS_ARM_MPU is always selected by SoC Kconfig. It is optional for the SoC; I wonder if you could reflect this in the SoC definition somehow.

Comment (henrikbrixandersen) [reply soc/arm/arm/designstart/soc.h:12]: I could introduce a SoC specific Kconfig for it and have that select CPU_HAS_ARM_MPU. Would that be better?

Comment (ioannisg) [reply soc/arm/arm/designstart/soc.h:12]: so, normally, it should be selected in: 
```
config SOC_ARM_DESIGNSTART_FPGA_CORTEX_M3
	bool "ARM Cortex-M3 DesignStart FPGA"
	select CPU_CORTEX_M3
	select XIP
	select HAS_SEGGER_RTT
```

but if you think that it is optional in this SoC, you can add a soc VARIANT and select it there. This is also done in several ARM SoC definitions.

Comment (henrikbrixandersen) [reply soc/arm/arm/designstart/soc.h:12]: I modeled this similar as to how you recommended handling the optional SysTick is handled for the Cortex-M1. Are you sure CPU_HAS_ARM_MPU should be handled differently?

Comment (ioannisg) [reply soc/arm/arm/designstart/soc.h:12]: No, it should not be handled differently. Just that it it starts getting weird with all these SoC-related options defined in Board level.

There's a good reason for this, though; this is an FPGA definition at the BOARD level. So we can live with this.

Comment (henrikbrixandersen) [reply soc/arm/arm/designstart/soc.h:12]: Okay. I will leave it as-is, then.

File: boards/arm/arty/doc/index.rst [hunk 2]
```diff
@@ -25,15 +25,18 @@ The Spartan-7 and Artix-7 based Arty board do not contain a CPU, but require a
...
 
-ARM Cortex-M1 DesignStart FPGA
-******************************
+ARM Cortex-M1/M3 DesignStart FPGA
+*********************************
 
-One such soft processor design is the Cortex-M1 `ARM DesignStart FPGA`_ Xilinx
-edition reference design from ARM. This design targets the Spartan-7 or Artix-7
-based Arty boards. Zephyr only supports the Artix-7 based boards for now.
+One way of instatiating soft processors on the Arty is using the `ARM
```
Comment (MaureenHelm) [inline boards/arm/arty/doc/index.rst:31]:
```suggestion
One way of instantiating soft processors on the Arty is using the `ARM
```

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Cool, thanks @henrikbrixandersen !

File: boards/arm/arty/doc/index.rst (hunk 2 above)
Comment (henrikbrixandersen) [reply boards/arm/arty/doc/index.rst]: Thanks @MaureenHelm. Fixed.

**Review by MaureenHelm — APPROVED** [review]

File: boards/arm/arty/Kconfig.board [hunk 3]
```diff
@@ -6,3 +6,8 @@
 config BOARD_ARTY_A7_ARM_DESIGNSTART_M1
 	bool "Digilent Arty A7 ARM DesignStart Cortex-M1"
 	depends on SOC_SERIES_ARM_DESIGNSTART
+
+config BOARD_ARTY_A7_ARM_DESIGNSTART_M3
+	bool "Digilent Arty A7 ARM DesignStart Cortex-M3"
+	select CPU_HAS_ARM_MPU
```
Comment (ioannisg) [inline boards/arm/arty/Kconfig.board:12]: @henrikbrixandersen 
Thinking more about this: would be more elegant if you move it in Kconfig.defconfig, under the definition of CPU_CORTEX_M_HAS_SYSTICK and make it default y if the M3 BOARD is used.

That's more for having both _HAS_ Kconfig options in one place; it should not change anything.

Comment (henrikbrixandersen) [reply boards/arm/arty/Kconfig.board]: @ioannisg Thanks, that's a bit more clean. Done.

**Review by ioannisg — APPROVED** [review]

**Pull request merged**